polar-core/src/sources.rs

```diff
@@ -1,17 +1,52 @@
+use std::{fmt, sync::Arc};
+
 use serde::{Deserialize, Serialize};
-use std::collections::HashMap;
 
-#[derive(Debug, Clone, Hash)]
+use crate::{formatting::source_lines, lexer::loc_to_pos};
+
+/// Parsed source context.
+#[derive(Clone, Debug)]
+pub struct Context {
+    pub source: Arc<Source>,
+    /// Start location within `source`.
+    pub left: usize,
+    /// End location within `source`.
+    pub right: usize,
+}
+
+impl fmt::Display for Context {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.source_position())?;
+        let lines = source_lines(&self.source, self.left, 0).replace('\n', "\n\t");
+        writeln!(f, ":\n\t{}", lines)?;
+        Ok(())
+    }
+}
+
+impl Context {
+    pub(crate) fn new(source: &Arc<Source>, left: usize, right: usize) -> Self {
+        Self {
+            source: source.clone(),
```

Comment:
Do the clone outside of this constructor and take an `Arc<Source>` by value. This way, the caller can choose whether to move the source into the `Context` (if it no longer needs a reference), or whether it needs to clone.

Reply:
f266fd51d